### PR TITLE
[Feature]. Improve DeltaWriter's performance.

### LINKED ISSUE #7214: [Feature] Improve DeltaWriter's performance
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

# Background

DeltaWriter may encounter serious performance problems when write concurrency is high, such as the following case
- There are a large number of insert-into-select tasks in the cluster
- When use BrokerLoad to load  large quantities of data with high concurrency

The following is a flame graph of a cluster which contains a large number of insert-into-select tasks

image

The following is a flame graph when  executing brokerload. The parallel of brokerload is set to 4, data to load: 200GB, 40 billion

image

# Solution

- When writing data to the DeltaWriter, group the data by tablet first and then write in batch, which can greatly reduce the lock/unlock operation

- Use mutex instead of SpinLock, because DeltaWriter.write is not a lightweight operation(It may hold the lock for a long time), using SpinLock while consumes too much CPU when the write concurrency is high

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by killxdcj**
## Proposed changes

Improve DeltaWriter's performance, detail in #7214

- Support batch write for DeltaWriter.

- Use mutex instead of SpinLock.

## Types of changes

What types of changes does your code introduce to Doris?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [x] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

- [x] I have created an issue on (Fix #7214) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

### CODE REVIEW

File: be/src/olap/delta_writer.cpp [hunk 1]
```diff
@@ -179,6 +180,26 @@ OLAPStatus DeltaWriter::write(Tuple* tuple) {
...
+        RETURN_NOT_OK(init());
+    }
+
+    if (_is_cancelled) {
+        return OLAP_ERR_ALREADY_CANCELLED;
+    }
+
+    for (const auto& row_idx : row_idxs) {
+        _mem_table->insert(row_batch->get_row(row_idx)->get_tuple(0));
+        if (_mem_table->memory_usage() >= config::write_buffer_size) {
```
Comment (morningman) [inline be/src/olap/delta_writer.cpp:195]: I think we can move this memory check outside the for loop.
Maybe inaccurate, but its ok~

Comment (killxdcj) [reply be/src/olap/delta_writer.cpp]: Comment resolved~

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by morningman** [comment]
Hi @killxdcj , please fix BE ut.

**Comment by killxdcj** [comment]
> 

@morningman BE ut fixed~

**Review by morningman — APPROVED** [review]

**Pull request merged**